### PR TITLE
New tests, refactored code

### PR DESCRIPTION
**Pull request by weirdGuy**
Particulary refactored code for message formatting, covered with tests

### CODE REVIEW

**Review by zelima — CHANGES_REQUESTED** [review]
@weirdGuy you've got the idea of having helper functions and test that functions individually, but think it's necessary to have separate function for getting each  part of log message.
So instead of repeating yourself and having separate `getAction()`, `getName()`, `getAssignees()` functions that each take same argument I would recommend to have one function Eg `getLogs()` that takes text and returns object. Something like `{"action": "+todo", timestamp: New date}`

File: scripts/main.coffee [hunk 1]
```diff
@@ -19,11 +19,11 @@ module.exports = (robot) ->
...
 
   robot.hear /.*/i, (res) ->
-      message = res.message.text.match(/\+[^*\s]+/)
+      message = messages.getAction(res.message.text)
       for key, val of config.monitor
           if(message != null)
-              if(key == message[0])
-                  for de in config.monitor[message[0]].dest
+              if(key == message)
+                  for de in config.monitor[message].dest
```
Comment (zelima) [inline scripts/main.coffee:23]: I would even pass the config to the function `getLogs(text, config)` to avoid iterating here. something like this Eg:
```
let getLogs(text, config) => {
  let monitor = Object.values(config.monitor)
  let act
  monitor.forEach(action) =>{
    act = text.find(action)
  }
  if (!act) {
    return
  }
  switch act:
    case '+todo':
      do something
   case '...'
   ....
}
```

Note this is just an example and suggestion. You may think of better approach. feel free to suggest if you have other ideas

File: scripts/messages.js [hunk 2]
```diff
@@ -112,11 +104,35 @@ let getRoom = function (roomId, callback) {
     })
 }
 
+let getAction = function (text) {
```
Comment (zelima) [inline scripts/messages.js]: Also can we stick to ES6 convention and use arrow functions

File: scripts/messages.js [hunk 3]
```diff
@@ -58,81 +58,40 @@ let sendMessage = function (message, dest, callback) {
...
-        callback(action.trim().substring(1) + ", " + new Date().toISOString() + ", " + "@" + message.user.login + " ("+ name +")" + ", " + roomid +", " + assignees + ", " + msg)
-    })
+let formatGist = (message, callback) => {
+    let action = formatting.getDataMask(message.text, /\+[^*\s]+/)
+    let assignees = formatting.getDataMask(message.text, /@[^*\s]+/)
+    let name = formatting.getName(message.user.name)
+    let msg = formatting.removeFromMessage(message.text, action)
 
+    formatting.getRoom(message.room.then(function (room) {
+        callback(action.substr(1) + ", " + new Date().toISOString() + ", " + "@" + message.user.login + " ("+ name +")" + ", " + room.name +", " + assignees + ", " + formatting.removeFromMessage(msg, assignees))
```
Comment (zelima) [inline scripts/messages.js]: Can we remove spaces after comas? Eg: have `","` instead of  `", "`

File: test/messages-test.js [hunk 4]
```diff
@@ -1,40 +1,48 @@
...
+        save = sinon.stub(new Gists(), 'download').callsFake(function (dest, callback) {
+            sinon.assert.calledOnce(save)
         })
 
-        messages.sendGist(message, "dest", function (inf) {})
+        messages.sendGist({"text": "+todo @test this", "user":{"name": "test", "login": "test"}}, "gdoc", function (data) {
 
-        sinon.assert.calledOnce(messages.sendGist)
+        })
+        save.reset()
```
Comment (zelima) [inline test/messages-test.js]: Again all this 3 test cases are useless. They are not actually testing anything. If you remove any of this calls from test cases it still passes
```
it('Gists Logging', function () {
        save = sinon.stub(new Gists(), 'download').callsFake(function (dest, callback) {
            sinon.assert.calledOnce(save)
        })
        save.reset()
    })
``` 
Eg if you add example code above to tests, it still passes

File: test/messages-test.js [hunk 5]
```diff
@@ -1,40 +1,48 @@
...
-        messages.sendMessage(message, "dest", function (inf) {})
+        messages.formatMessage({"text": "+todo @test this", "room": "testingName", "user": {"name": "test", "login": "test"}}, function (data) {
+        })
+        save.reset()
+    })
 
-        sinon.assert.calledOnce(messages.sendMessage)
+    it('GDoc Logging', function () {
+        save = sinon.stub(messages, 'formatMessage').callsFake(function (message, callback) {
+            sinon.assert.calledOnce(save)
```
Comment (zelima) [inline test/messages-test.js]: Also besides `calledOnce` we wan't to know weather it was called with correct message
See `sinon.spy.calledWith(message)` for example

File: test/parsing-test.js [hunk 6]
```diff
@@ -21,49 +22,35 @@ describe('Messages parsing', function () {
     })
 
     it('Action getting', function () {
-        assert.equal(messages.getAction(msg), "+todo")
+        assert.equal(formatting.getDataMask(msg, /\+[^*\s]+/), "+todo")
```
Comment (zelima) [inline test/parsing-test.js]: We are testing against one fixed message, that is exacted behavior from user. What about unexpected behaviors?? Eg:
- will have typo in `+tod` instead `+todo`
- what if not `+todo` at all, is it still logging?
- what if 2 assignee in one message Eg +todo review code @weirdGuy @zelima

Also 
- what if +todo is in the middle, at the end?
- what if no assignee 
etc... 

We need test cases for above as well

**Review by zelima — CHANGES_REQUESTED** [review]
Much better, some more changes and we are almost there :)

File: test/parsing-test.js (hunk 6 above)
Comment (weirdGuy) [reply test/parsing-test.js]: If in command typo mistake it won't be executed at all on the stage where we are looking for action

Comment (zelima) [reply test/parsing-test.js]: @weirdGuy I know, just we need tests for that. To be sure that it is actually happening in that way ;)

File: scripts/messages.js [hunk 7]
```diff
@@ -49,7 +49,8 @@ let sendMessage = (message, dest, callback) => {
         doc = new GoogleSpreadsheet(dest)
         setAuth(function (err, info) {
             if(!err) {
-                doc.addRow(worksheetinfo.worksheets[0].id, res, function (err, info) {
+                console.log(doc)
```
Comment (zelima) [inline scripts/messages.js]: why do wee need to log this?

File: test/parsing-test.js [hunk 8]
```diff
@@ -12,13 +12,20 @@ let helper = new Helper('../scripts/main.coffee')
 
 describe('Messages parsing', function () {
     let msg
+    let room
+    let a
+    let b
```
Comment (zelima) [inline test/parsing-test.js]: `a` and `b`?? not really  getting what are these - Gdocs and  Gist? can we have more reasonable names for variables?

File: test/messages-test.js [hunk 9]
```diff
@@ -1,48 +1,88 @@
...
+    //
+    // it('Gists Logging', function () {
+    //     save = sinon.stub(new Gists(), 'download').callsFake(function (dest, callback) {
+    //         sinon.assert.calledOnce(save)
+    //     })
+    //
+    //     messages.sendGist({"text": "+todo @test this", "user":{"name": "test", "login": "test"}}, "gdoc", function (data) {
+    //
+    //     })
+    // })
```
Comment (zelima) [inline test/messages-test.js]: Why committing file with commented everything init  ??

File: scripts/main.coffee [hunk 10]
```diff
@@ -13,13 +13,14 @@ module.exports = (robot) ->
...
   messages = require('./messages.js').messages
+  formatting = require('./formatting.js').formatting
 
   # messages.setAuth((err, res) ->
   #     console.log("Logged in: " + res[1].title)
   # )
 
   robot.hear /.*/i, (res) ->
-      message = messages.getAction(res.message.text)
+      message = formatting.getDataMask(res.message.text, /\+[^*\s]+/)
```
Comment (zelima) [inline scripts/main.coffee]: Is this line related to tests? why have you changed the name, found the bug? if so I suggest doing this in separate commit, cause it is really confusing.

File: test/parsing-test.js [hunk 11]
```diff
@@ -47,10 +54,67 @@ describe('Messages parsing', function () {
     })
 
     it('Message Formatting', function () {
-        sinon.stub(formatting, 'getRoom').resolves({"name": "test"})
+        let t = sinon.stub(formatting, 'getRoom').resolves({"name": "test"})
```
Comment (zelima) [inline test/parsing-test.js]: again `t` ?

**Review by zelima — CHANGES_REQUESTED** [review]
Think parsin-test.js looks good - small comments re variable naming.
Waiting for message-test.js refactor

@weirdGuy also please write reasonable commit messages and reference to the issues. `Test improvements` is very general. I suggest summarizing your work in you commit messages. Eg: 
"[tests][m]: new tests against unexpected user behavior -refs # isseu number"

File: scripts/messages.js (hunk 7 above)
Comment (weirdGuy) [reply scripts/messages.js]: @zelima Forgot to delete from testing :)

File: scripts/main.coffee (hunk 10 above)
Comment (weirdGuy) [reply scripts/main.coffee]: @zelima i have moved part of methods from messages, that are just refferr to formatting messages, not to just sending them

Comment (zelima) [reply scripts/main.coffee]: Ok, but why it is the part of this commit? i do not see any changes to `messages` in this commit. 

I would do this in separate commit with reasonable message Eg "moved part of methods from messages, that are just reffer to formatting messages, not to just sending them"

File: scripts/messages.js [hunk 12]
```diff
@@ -30,7 +30,7 @@ let sendMessage = (message, dest, callback) => {
     formatMessage(message, function (res) {
         gdocs.setAuth(dest, function (err, info) {
             if(!err) {
-                gdocs.addRow(info.worksheets[0].id, res, function (err, info) {
+                gdocs.addRow(info[1].worksheets[0].id, res, function (err, info) {
```
Comment (zelima) [inline scripts/messages.js]: Not really getting this one `info[1]` - is this always the case? what if there is not element 1? will not this `error`?

**Review by zelima — APPROVED** [review]

**Comment by weirdGuy** [comment]
Changes:
- Refactored methods inside ```messages.js```
- Created and improved methods for messages formatting and messages sending
- Methods for working with GDocs were moved to ```gdocs.js```
- Methods to format messages were moved to ```formatting.js```

**Comment by weirdGuy** [comment]
Tests:
```messages-test.js```:
- Gist sending
- GDocs sending

```parsing-test.js```:
- Action getting
- Action removing from message
- Assignees getting
- Assignees removing from message
- Action and assignees removing from message
- Name getting
- Message formatting
- Message formatting with tag in middle/end
- Message with typo in tag
- Gist formatting
- GDoc formatting
- Calling/not calling send function with tag typo, without tag, tag in different places of message

File: scripts/messages.js (hunk 12 above)
Comment (weirdGuy) [reply scripts/messages.js]: This one item always in this case, it returns smthing like ```(err, info)``` and ```info``` always look's like:
```
info[null,{worksheets: []...}]
```
